torch/_higher_order_ops/while_loop.py

```diff
@@ -0,0 +1,126 @@
+    track_tensor_tree,
+)
+
+
+while_loop_op = HigherOrderOperator("while_loop")
+
+
+def while_loop(cond_fn, body_fn, operands):
+    # TODO(yidi): turn on dynamo in eager-mode
+    assert isinstance(operands, (tuple, list))
+    return while_loop_op(cond_fn, body_fn, operands)
+
+
+@while_loop_op.py_impl(DispatchKey.CompositeExplicitAutograd)
+def while_loop_dense(cond_fn, body_fn, operands):
+    init_val = operands
+    while cond_fn(*init_val):
+        out = body_fn(*init_val)
+        assert isinstance(
+            out, (tuple, list)
+        ), "body_fn should return the same type as operands"
+        assert len(out) == len(
+            init_val
+        ), "body_fn should return the same number of elements as operands"
+        init_val = out
+    return init_val
+
+
+while_loop_op.py_impl(DispatchKey.Autograd)(
+    autograd_not_implemented(while_loop_op, deferred_error=True)
+)
+
+
+@while_loop_op.py_impl(ProxyTorchDispatchMode)
+def while_loop_tracing(mode, cond_fn, body_fn, operands):
+    def _trace_while_loop(proxy_mode, while_loop_op, cond_fn, body_fn, operands):
+        pre_dispatch = getattr(proxy_mode, "pre_dispatch", False)
+
+        with disable_proxy_modes_tracing():
+            cond_graph = make_fx(
+                _maybe_run_with_interpreter(cond_fn), pre_dispatch=pre_dispatch
+            )(*operands)
+            body_graph = make_fx(
+                _maybe_run_with_interpreter(body_fn), pre_dispatch=pre_dispatch
+            )(*operands)
+
+        next_name = None
+        i = 0
+        while not next_name:
+            candidate = f"while_loop_cond_graph_{i}"
+            if hasattr(proxy_mode.tracer.root, candidate):
+                i += 1
+            else:
+                next_name = candidate
+        cond_graph_name = next_name
+        body_graph_name = f"while_loop_body_graph_{i}"
+        assert not hasattr(proxy_mode.tracer.root, body_graph_name)
+
+        proxy_mode.tracer.root.register_module(cond_graph_name, cond_graph)
+        proxy_mode.tracer.root.register_module(body_graph_name, body_graph)
+
+        args = (cond_graph, body_graph, operands)
+
+        proxy_args = pytree.tree_map(proxy_mode.tracer.unwrap_proxy, args)
+
+        out_proxy = proxy_mode.tracer.create_proxy(
+            "call_function", while_loop_op, proxy_args, {}, name="while_loop"
+        )
+
+        # body_fn return output with the same pytree and tensor meta data as operands
+        # so we could just return the output after one iteration.
+        out = body_fn(*operands)
+        return track_tensor_tree(
+            out, out_proxy, constant=None, tracer=proxy_mode.tracer
+        )
+
+    if mode.enable_tracing:
+        return _trace_while_loop(mode, while_loop_op, cond_fn, body_fn, operands)
+    else:
+        return while_loop_op(cond_fn, body_fn, operands)
+
+
+@while_loop_op.py_impl(FakeTensorMode)
+def while_loop_fake_tensor_mode(mode, cond_fn, body_fn, operands):
+    return body_fn(*operands)
```

Comment:
should probably be while_loop_dense, the number of iterations may actually matter

Reply:
We cannot use while_loop_dense here because cond_fn returns a fake tensor and while cond_fn() will need to actually get the value out of the boolean scalar tensor in order to do the branching, which gives us a data dependent error: `trying to get a value out of symbolic int/float whose value is data-dependent (and thus we do not know the true value.)  The expression we were trying to evaluate is Eq(i0, 1) (unhinted: Eq(i0, 1)).`

We may have to assume different iterations will return the same shaped tensor and use the body_fn's return as the output fake tensor shape (or just clone the operands?).

Reply:
Okay, let's stick with this for now